src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -1107,4 +1113,181 @@ public LeafRope flatten(Rope rope) {
+        }
+
+        public abstract boolean execute(Rope a, Rope b);
+
+        @Specialization(guards = "a == b")
+        public boolean sameRopeEqual(Rope a, Rope b) {
+            return true;
+        }
+
+        @Specialization
+        public boolean ropesEqual(Rope a, Rope b,
+                @Cached("create()") BytesNode aBytes,
+                @Cached("create()") BytesNode bBytes) {
+            return a.hashesMatch(b) && a.byteLength() == b.byteLength() && Arrays.equals(aBytes.execute(a), bBytes.execute(b));
+        }
+
+        @Specialization
+        public boolean ropeEqualNonRope(Rope a, Object b) {
+            return false;
+        }
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "rope")
+    })
+    public abstract static class BytesNode extends RubyNode {
+
+        public static BytesNode create() {
+            return RopeNodesFactory.BytesNodeGen.create(null);
+        }
+
+        public abstract byte[] execute(Rope rope);
+
+        @Specialization(guards = "rope.getRawBytes() != null")
+        public byte[] getBytesOnHeap(OnHeapRope rope) {
+            return rope.getRawBytes();
+        }
+
+        public byte[] getBytesNative(NativeRope rope) {
+            return rope.getBytes();
+        }
+
+        @Specialization(guards = "rope.getRawBytes() == null")
+        @TruffleBoundary
+        public byte[] getBytesFromRope(OnHeapRope rope) {
+            return rope.getBytes();
+        }
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "rope")
+    })
+    public abstract static class BytesSlowNode extends RubyNode {
+
+        public static BytesSlowNode create() {
+            return RopeNodesFactory.BytesSlowNodeGen.create(null);
+        }
+
+        public abstract byte[] execute(Rope rope);
+
+        @Specialization
+        public byte[] getNativeBytesSlow(NativeRope rope) {
+            return rope.getBytes();
+        }
+
+        @Specialization
+        public byte[] getNativeBytesSlow(RepeatingRope rope) {
+            return rope.getBytesSlow();
+        }
+
+        @Specialization
+        public byte[] getNativeBytesSlow(SubstringRope rope) {
+            return rope.getBytesSlow();
+        }
+
+        @Specialization
+        public byte[] getBytesFromRope(Rope rope) {
+            return RopeOperations.flattenBytes(rope);
+        }
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "rope"),
+            @NodeChild(type = RubyNode.class, value = "index")
+    })
+    public abstract static class ByteSlowNode extends RubyNode {
+
+    public static ByteSlowNode create() {
+            return RopeNodesFactory.ByteSlowNodeGen.create(null, null);
+        }
+
+        public abstract byte execute(Rope rope, int index);
+
+        @Specialization
+        public byte getByteFromSubString(SubstringRope rope,
+                int index,
```

Comment:
The `index` should be on the same line as `rope`, only `@Cached` parameters have their own line.